qiskit/algorithms/optimizers/adam_amsgrad.py

```diff
@@ -114,6 +114,21 @@ def __init__(
                 writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
                 writer.writeheader()
 
+    def to_dict(self) -> Dict[str, Any]:
```

Comment:
I updated `from_dict` to handle all cases in the base class. I thought this might be tricky since I wanted to allow users to omit the "name" if `from_dict` is called on a subclass (e.g. SPSA), but turns out it's not 👍🏻 

Concerning `to_dict`, I'm not sure if there's enough common structure to implement that in the base class. I'd be adding special implementations to probably half the classes anyways, or what do you think?